domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -397,110 +437,31 @@ class ExplorationProgressController @Inject constructor(
-    if (!explorationProgress.shouldSavePartialProgress) return
-    val profileId: ProfileId = explorationProgress.currentProfileId
-    val topicId: String = explorationProgress.currentTopicId
-    val storyId: String = explorationProgress.currentStoryId
-    val explorationId: String = explorationProgress.currentExplorationId
-
-    val checkpoint: ExplorationCheckpoint =
-      explorationProgress.stateDeck.createExplorationCheckpoint(
-        explorationProgress.currentExploration.version,
-        explorationProgress.currentExploration.title,
-        oppiaClock.getCurrentTimeMs()
-      )
-
-    val deferred = explorationCheckpointController.recordExplorationCheckpointAsync(
-      profileId,
-      explorationId,
-      checkpoint
-    )
-
-    deferred.invokeOnCompletion {
-      val checkpointState = if (it == null) {
-        deferred.getCompleted()
-      } else {
-        oppiaLogger.e("Lightweight checkpointing", "Failed to save checkpoint in exploration", it)
-        // CheckpointState is marked as CHECKPOINT_UNSAVED because the deferred did not
-        // complete successfully.
-        CheckpointState.CHECKPOINT_UNSAVED
-      }
-      processSaveCheckpointResult(
-        profileId,
-        topicId,
-        storyId,
-        explorationId,
-        oppiaClock.getCurrentTimeMs(),
-        checkpointState
-      )
+  /** Stops any new hints and solution from showing up. */
+  fun stopNewHintsAndSolutionFromShowingUp() {
+    explorationProgressLock.withLock {
+      hintHandler.hideHintsAndSolution()
     }
   }
 
   /**
-   * Processes the result obtained upon complete execution of the function
-   * [saveExplorationCheckpoint].
+   * Notifies the [HintHandler] that a scheduled task to show new help has finished.
    *
-   * Marks the exploration as in_progress_saved or in_progress_not_saved if it is not already marked
-   * correctly. This function also updates the checkpoint state of the exploration to the
-   * specified new checkpoint state.
-   *
-   * @param profileId is the profile id currently playing the exploration
-   * @param topicId is the id of the topic which contains the story with the current exploration
-   * @param storyId is the id of the story which contains the current exploration
-   * @param lastPlayedTimestamp timestamp of the time when the checkpoints state for the exploration
-   *     was last updated
-   * @param newCheckpointState the latest state obtained after saving checkpoint successfully or
-   *     unsuccessfully
+   * @param trackedSequenceNumber the ID used to identify each task scheduled to show help
+   * @param state the state on which the hint is shown
    */
-  private fun processSaveCheckpointResult(
-    profileId: ProfileId,
-    topicId: String,
-    storyId: String,
-    explorationId: String,
-    lastPlayedTimestamp: Long,
-    newCheckpointState: CheckpointState
-  ) {
+  fun hintAndSolutionTimerCompleted(trackedSequenceNumber: Int, state: State) {
```

Comment:
I am not entirely sure how we will move the timer to the domain layer, more specifically how we will stop and resume it when the activity resumes or pauses. We can use the override methods onPause and onResume but how will we notify the timer to pause.

I looked into the default timer that android has,
```
 object : CountDownTimer(dealyToShowHints, 1000) {

     override fun onTick(millisUntilFinished: Long) {}

     override fun onFinish() {
         // reveal hint or solution 
     }
 }.start()
 
```



 Alternatively, we can also use coroutine,
```
    backgroundCoroutineScope.launch {
      delay(delayToShowHints)
      // reveal hint/solution
    }
```

But in either of the above methods, I am not sure how we will pause the timer and resume it on activity pause or resume. 

One possible way to handle pause and resume cases can be to save the delay left and cancel the timer on pause left and on resume rerun a new timer with the remaining delay.

```
 object : CountDownTimer(dealyToShowHints, 1000) {

     override fun onTick(millisUntilFinished: Long) {
          if (hintSequenceNumber == mostRecentSequenceNumber) {
               explorationProgress.timeEllapsed += 1
         }
     }

     override fun onFinish() {
         // reveal hint or solution 
     }
 }.start()
```
But I think we can do it like this only with the default timers and not with coroutines.

Can you please suggest if one of the above mentioned ways are correct or give a few points on how I should approach this problem.

Reply:
Hmm. I'm not worried about the locking mechanism (yet) since the functions return relatively quickly.

Regarding how to do this, I think there are a couple ways we can do this that will match the current behavior. However, it occurs to me that the current behavior may be wrong (I hypothesize that if you were to pause the app after triggering the timer & wait for the timer's duration before returning, a hint should show up since the timer continues while the app is backgrounded).

For the sake of time, I suggest we go with a parity approach and file an issue tracking this bug since it'll require more work to solve. Could you please file that issue?

Regarding the solution, I don't think we actually need to do anything extra. The timer will continue while the app is backgrounded, recompute the HelpIndex & EphemeralState, notify subscribers, and post the data to the UI. Fortunately, that last part happens via LiveData in the same way the lifecycle-safe timer does, so the result shouldn't actually be observed until the user returns. I think the current batch of tests should cover these cases well enough that if they pass, the behavior should be working as expected with the timer moved to the domain layer.

Reply:
moved the timer to the domain layer.